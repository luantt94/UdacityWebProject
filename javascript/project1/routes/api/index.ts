import express, { Router, Request, Response } from "express";
import sharp from "sharp";
import fs from "fs";
// import sharp from "sharp";

const router: Router = express.Router();

router.get("/", async (req: Request, res: Response) => {
  const { filename, width, height } = req.query;
  const newImgPath = `disk\\${filename}_${width}_${height}.jpg`;
  const folderPath = "./disk";
  let isExistedImg = false;

  const names = await fs.promises.readdir(folderPath);
  names.forEach((file) => {
    if (file.split(".")[0] === filename) {
      isExistedImg = true;
      return;
    }
  });

  if (!isExistedImg) {
    await sharp(`images\\${filename}.jpg`)
      .rotate()
      .resize({ width: Number(width), height: Number(height) })
      .toFile(newImgPath, (err, info) => {
        console.log(info);
      })
      .toBuffer();
  }

  const image = fs.readFileSync(newImgPath);
  res.setHeader("Content-Type", "image/jpeg");
  res.setHeader("Content-Length", image.length);

  res.write(image);
  res.end();
});

export default router;
