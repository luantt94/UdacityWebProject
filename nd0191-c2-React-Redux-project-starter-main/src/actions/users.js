export const RECEIVE_USERS = "RECEIVE_USERS";
export const USER_VOTE_ANSWER = "USER_VOTE_ANSWER";
export const INSERT_QID_TO_AUTHOR = "INSERT_QID_TO_AUTHOR";

export function receiveUsers(users) {
  return {
    type: RECEIVE_USERS,
    users,
  };
}

export function userVoteAnswer(authedUser, qid, answer) {
  return {
    type: USER_VOTE_ANSWER,
    authedUser,
    qid,
    answer,
  };
}

export function insertQIdToAuthor(authedUser, qid) {
  return {
    type: INSERT_QID_TO_AUTHOR,
    authedUser,
    qid,
  };
}
