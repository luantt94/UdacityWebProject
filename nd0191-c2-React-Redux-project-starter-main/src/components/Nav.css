.nav_b {
  display: flex;
  gap: 20px;
  /* margin-top: 30px; */
}

li {
  list-style-type: none;
}

.nav1 {
  display: flex;
  justify-content: space-between;
  margin-top: 30px;
}
.nav2 {
  display: flex;
  gap: 20px;
}
