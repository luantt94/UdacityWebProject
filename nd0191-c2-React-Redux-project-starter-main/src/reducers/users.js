import {
  RECEIVE_USERS,
  USER_VOTE_ANSWER,
  INSERT_QID_TO_AUTHOR,
} from "../actions/users";

export default function users(state = {}, action) {
  switch (action.type) {
    case RECEIVE_USERS:
      return {
        ...state,
        ...action.users,
      };
    case USER_VOTE_ANSWER:
      return {
        ...state,
        [action.authedUser]: {
          ...state[action.authedUser],
          answers: {
            ...state[action.authedUser].answers,
            [action.qid]: action.answer,
          },
        },
      };
    case INSERT_QID_TO_AUTHOR:
      return {
        ...state,
        [action.authedUser]: {
          ...state[action.authedUser],
          questions: state[action.authedUser].questions.concat(action.qid),
        },
      };
    default:
      return state;
  }
}
