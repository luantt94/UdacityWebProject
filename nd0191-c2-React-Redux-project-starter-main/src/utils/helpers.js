export function formatDate(timestamp) {
  const d = new Date(timestamp);
  const time = d.toLocaleTimeString("en-US");
  return time.substr(0, 5) + time.slice(-2) + " | " + d.toLocaleDateString();
}

export function formatQuestion(question) {
  const { id, author, timestamp } = question;

  return {
    id,
    author,
    timestamp,
  };
}

export function getQuestionStatus(questions, authedUser, questionId) {
  const isDone =
    questions[questionId].optionOne.votes.includes(authedUser) ||
    questions[questionId].optionTwo.votes.includes(authedUser);
  return {
    isDone,
    author: questions[questionId].author,
    optionOne: questions[questionId].optionOne.text,
    optionTwo: questions[questionId].optionTwo.text,
  };
}

export function getDoneQuestion(questions, authedUser) {
  return Object.values(questions).filter(
    (question) =>
      question.optionOne.votes.includes(authedUser) ||
      question.optionTwo.votes.includes(authedUser)
  );
}

export function getNotDoneQuestion(questions, authedUser) {
  return Object.values(questions).filter(
    (question) =>
      !question.optionOne.votes.includes(authedUser) &&
      !question.optionTwo.votes.includes(authedUser)
  );
}
