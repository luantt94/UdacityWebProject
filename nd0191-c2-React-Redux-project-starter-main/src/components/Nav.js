import { Link } from "react-router-dom";
import "./Nav.css";
const Nav = () => {
  return (
    <nav className="nav">
      <ul className="nav_b">
        <li>
          <Link to="/">Home</Link>
        </li>
        <li>
          <Link to="/leaderboard">Leaderboard</Link>
        </li>
        <li>
          <Link to="/new">New</Link>
        </li>
      </ul>
    </nav>
  );
};

export default Nav;
