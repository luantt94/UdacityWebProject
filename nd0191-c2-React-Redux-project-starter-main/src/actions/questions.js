import questions from "../reducers/questions";
import { saveQuestion, saveQuestionAnswer } from "../utils/api";
import { showLoading, hideLoading } from "react-redux-loading-bar";
import { insertQIdToAuthor } from "./users";
export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const ADD_QUESTION = "ADD_QUESTION";
export const VOTE_QUESTION_ANSWER = "VOTE_QUESTION_ANSWER";

function addQuestion(question) {
  return {
    type: ADD_QUESTION,
    question,
  };
}

function voteQuestionAnswer({ authedUser, qid, answer }) {
  return {
    type: VOTE_QUESTION_ANSWER,
    authedUser,
    qid,
    answer,
  };
}

export function handleVoteOption(authedUser, qid, answer) {
  saveQuestionAnswer({ authedUser, qid, answer });
  return (dispatch) => {
    return dispatch(voteQuestionAnswer({ authedUser, qid, answer }));
  };

  // return (dispatch) => {
  //   dispatch(showLoading());

  //   return saveQuestionAnswer({ authedUser, qid, answer })
  //     .then((question) =>
  //       dispatch(voteQuestionAnswer({ authedUser, qid, answer }))
  //     )
  //     .then(() => dispatch(hideLoading()));
  // };
}

export function handleAddQuestion(values) {
  return (dispatch, getState) => {
    const { authedUser } = getState();

    dispatch(showLoading());
    return saveQuestion({
      optionOneText: values.optionOneText,
      optionTwoText: values.optionTwoText,
      author: authedUser,
    })
      .then((question) => {
        dispatch(addQuestion(question));
        dispatch(insertQIdToAuthor(question.author, question.id));
      })
      .then(() => dispatch(hideLoading()));
  };
}

export function receiveQuestions(questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}
