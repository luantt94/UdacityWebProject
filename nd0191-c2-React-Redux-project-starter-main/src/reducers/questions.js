import { RECEIVE_QUESTIONS, VOTE_QUESTION_ANSWER } from "../actions/questions";

export default function questions(state = {}, action) {
  switch (action.type) {
    case RECEIVE_QUESTIONS:
      return {
        ...state,
        ...action.questions,
      };
    case VOTE_QUESTION_ANSWER:
      console.log("Stateeeeee");
      console.log(state);
      console.log("Actionnnnnnnn");
      console.log(action);
      // const s = {
      //   ...state,
      //   [action.qid]: {
      //     ...state[action.qid],
      //     [action.answer]: {
      //       ...state[action.qid][action.answer],
      //       votes: state[action.qid][action.answer].votes.concat([
      //         action.authedUser,
      //       ]),
      //     },
      //   },
      // };
      // console.log("ssssssss");
      // console.log(s);

      return {
        ...state,
        [action.qid]: {
          ...state[action.qid],
          [action.answer]: {
            ...state[action.qid][action.answer],
            votes: state[action.qid][action.answer].votes.concat([
              action.authedUser,
            ]),
          },
        },
      };
    default:
      return state;
  }
}
