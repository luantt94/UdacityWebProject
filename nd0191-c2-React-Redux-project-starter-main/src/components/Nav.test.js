import * as React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import Nav from "./Nav";
import middleware from "../middleware";
import reducer from "../reducers";
import { BrowserRouter } from "react-router-dom";
import { createStore } from "redux";
import { getInitialData } from "../utils/api";
import { act } from "react-dom/test-utils";
import { receiveUsers } from "../actions/users";
import { receiveQuestions } from "../actions/questions";
import { setAuthedUser } from "../actions/authedUser";

const store = createStore(reducer, middleware);
describe("Nav", () => {
  beforeAll(async () => {
    await getInitialData().then((data) =>
      act(() => {
        store.dispatch(receiveUsers(data.users));
        store.dispatch(receiveQuestions(data.questions));
        store.dispatch(setAuthedUser("mtsamis"));
      })
    );
  });
  it("will match snapshot", () => {
    const component = render(
      <BrowserRouter>
        <Provider store={store}>
          <Nav />
        </Provider>
      </BrowserRouter>
    );
    expect(component).toMatchSnapshot();
  });
});
