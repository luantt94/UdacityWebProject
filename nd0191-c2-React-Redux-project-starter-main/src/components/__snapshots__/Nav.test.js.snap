// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav will match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="nav1 "
      >
        <div>
          <ul
            class="nav_b"
          >
            <li>
              <a
                href="/"
              >
                Home
              </a>
            </li>
            <li>
              <a
                href="/leaderboard"
              >
                Leaderboard
              </a>
            </li>
            <li>
              <a
                href="/new"
              >
                New
              </a>
            </li>
          </ul>
        </div>
        <div>
          <ul
            class="nav2"
          >
            <li>
              <img
                alt=""
                height="60"
                src="3.PNG"
                width="60"
              />
              Mike Tsamis
            </li>
            <li>
              <button>
                logout
              </button>
            </li>
          </ul>
        </div>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="nav1 "
    >
      <div>
        <ul
          class="nav_b"
        >
          <li>
            <a
              href="/"
            >
              Home
            </a>
          </li>
          <li>
            <a
              href="/leaderboard"
            >
              Leaderboard
            </a>
          </li>
          <li>
            <a
              href="/new"
            >
              New
            </a>
          </li>
        </ul>
      </div>
      <div>
        <ul
          class="nav2"
        >
          <li>
            <img
              alt=""
              height="60"
              src="3.PNG"
              width="60"
            />
            Mike Tsamis
          </li>
          <li>
            <button>
              logout
            </button>
          </li>
        </ul>
      </div>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
